controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -170,11 +171,12 @@ public String getScopeName() {
     private CompletableFuture<StreamTruncationRecord> computeTruncationRecord(StreamTruncationRecord truncationRecord,
                                                                               Map<Integer, Long> streamCut) {
         log.debug("computing truncation for stream {}/{}", scope, name);
-        return getHistoryTableFromStore()
-                .thenCompose(history -> getSegmentTableFromStore()
-                        .thenCompose(segment -> getIndexTable()
-                                .thenApply(index -> TableHelper.computeTruncationRecord(index.getData(), history.getData(),
-                                        segment.getData(), streamCut, truncationRecord))));
+        return getHistoryIndexFromStore()
+                .thenCompose(historyIndex -> getHistoryTableFromStore()
```

Comment:
I think it is better to align the beginning of the lines rather than cascading like this.

Reply:
This is because thenCompose is being called on the method call inside thenCompose block. 
This is done so that the value from each future is available at all those below it.